storages/request_dispatcher.go

```diff
@@ -14,60 +21,116 @@ type dispatcher interface {
+	pickClientFactory         func(*http.Request) func([]*backend.Backend, watchdog.ConsistencyWatchdog) client
+	pickResponsePickerFactory func(*Request) func(<-chan BackendResponse) responsePicker
+	watchdog                  watchdog.ConsistencyWatchdog
+	watchdogRecordFactory     watchdog.ConsistencyRecordFactory
 }
 
 // NewRequestDispatcher creates RequestDispatcher instance
-func NewRequestDispatcher(backends []*backend.Backend, syncLog *SyncSender) *RequestDispatcher {
-
+func NewRequestDispatcher(backends []*backend.Backend, syncLog *SyncSender,
+	watchdog watchdog.ConsistencyWatchdog, watchdogRecordFactory watchdog.ConsistencyRecordFactory) *RequestDispatcher {
 	return &RequestDispatcher{
 		Backends:                  backends,
 		syncLog:                   syncLog,
 		pickResponsePickerFactory: defaultResponsePickerFactory,
 		pickClientFactory:         defaultReplicationClientFactory,
+		watchdog:                  watchdog,
+		watchdogRecordFactory:     watchdogRecordFactory,
 	}
 }
 
 // Dispatch creates and calls replicators and response pickers
 func (rd *RequestDispatcher) Dispatch(request *http.Request) (*http.Response, error) {
+	storageRequest := &Request{request, nil, nil}
+	if rd.shouldUseConsistencyWatchdog(request) {
+		recordedRequest, err := rd.createAndInsertRecordFor(request)
+		if err != nil {
+			return nil, err
+		}
+		storageRequest.record = recordedRequest.record
+		storageRequest.marker = recordedRequest.marker
+	}
 	clientFactory := rd.pickClientFactory(request)
-	cli := clientFactory(rd.Backends)
-	respChan := cli.Do(request)
-	pickerFactory := rd.pickResponsePickerFactory(request)
+	cli := clientFactory(rd.Backends, rd.watchdog)
+	respChan := cli.Do(storageRequest)
+	pickerFactory := rd.pickResponsePickerFactory(storageRequest)
 	pickr := pickerFactory(respChan)
 	go pickr.SendSyncLog(rd.syncLog)
 	return pickr.Pick()
 }
+func (rd *RequestDispatcher) createAndInsertRecordFor(request *http.Request) (*Request, error) {
+	record, err := rd.watchdogRecordFactory.CreateRecordFor(request)
+	if err != nil {
+		return nil, err
+	}
+	if isInitiateMultiPartUploadRequest(request) {
+		record.ExecutionDate = record.ExecutionDate.Add(oneWeek).Add(time.Minute * 10)
+	}
+	deleteMarker, err := rd.watchdog.Insert(record)
+	if err != nil {
+		return nil, err
+	}
+	return &Request{
+		request,
+		record,
+		deleteMarker,
+	}, nil
+}
+func (rd *RequestDispatcher) shouldUseConsistencyWatchdog(request *http.Request) bool {
+	isMultiPartRequest := isMultiPartUploadRequest(request)
+	isInitiateMultiPart := isInitiateMultiPartUploadRequest(request)
+
+	consistencyCondition := rd.watchdog != nil && len(rd.Backends) > 1
+
+	methodCondition := (http.MethodPut == request.Method && !isMultiPartRequest) ||
+						http.MethodDelete == request.Method ||
+						(http.MethodPost == request.Method && isInitiateMultiPart)
+
+	pathCondition := !utils.IsBucketPath(request.URL.Path)
```

Comment:
What if we would skip this condition? Maybe watchdog & brim could fix client bucket listings on storage failure?